docs/jj_first_principles.md

```diff
@@ -0,0 +1,50 @@
+# Jujutsu from first principles
+
+## Preface
+
+Why does Jujutsu exist and which problems does it solve? This document tries to
+answer both of these questions while expanding on the design in a user-friendly
+way. 
+
+At its core Jujutsu is [Version Control System][vcs] which scales to huge 
+repositories at [Google scale][billion-lines]. Many design choices are 
+influenced by the concurrent commits happening in Googles Monorepo, as there 
+are always multiple people working on the same file(s) at the same time.
+
+## Core Tenets
+
+Jujutsu's core tenets are:
+
+ 1. Having as few states as possible.
```

Comment:
> These principles are more or less consistent with what `jj` is now, but I'm not sure they are good enough to be stated as "core tenets" of `jj`. They are a bit vague and I'm worried it pins our down a bit too much. I also feel differently about them, e.g. I'm not sure I believe in "having as few states as possible" even though, all else being equal, fewer states is of course better than more.

Fair enough, I do think that the native backend should strive to keep it that way. 

> We might want to add more states at some points (e.g. tracked and untracked branches were added recently, we might want some concept of topics).

Here I tend do disagree with you, the whole tracked/untracked branch state only got introduced for Git interop, which isn't a concern for this doc. And topics are pure commit metadata, not another state, the main change will be in the clients.  

> We can never make it so that it's "fundamentally impossible' to lose work in your repository, and while I believe quite a bit in trying to make that hard, a large increase in convenience might be worth a potential footgun.

While that is true, some backend implementations can make if fundamentally impossible, e.g the Google Spanner one.  So this is should be a implementation detail which isn't surfaced on the CLI. So we won't trade convenience there.

Reply:
This is tangentially related, but it sounds to me that one of your goals here is to describe what JJ would look like without a git backend. Perhaps we could make it clearer in the doc. We could rename all of it to "JJ without Git" or "JJ from First Principles (without Git)", or split out sections that apply more to the Git-backed `jj` or to Git-less `jj`. I think I missed that this was your focus when I was first looking this over.

BTW, I may sound a bit conflicted on this topic, so I'll write the following down: I absolutely agree that it is important that `jj` makes sense conceptually without Git, so this is very much worth thinking about. At the same time, I personally think that the git backend is an absolutely essential part of `jj` and will continue to be so for years while people hopefully adopt it. However, I don't think everyone agrees with this, and I also hope that it will slowly become less essential (though I do imagine this change to be slow for a while).

Reply:
> This is tangentially related, but it sounds to me that one of your goals here is to describe what JJ would look like without a git backend. Perhaps we could make it clearer in the doc. We could rename all of it to "JJ without Git" or "JJ from First Principles (without Git)", or split out sections that apply more to the Git-backed `jj` or to Git-less `jj`.

Yes, that's correct, so I've adjusted the title to your second suggestion. 


> Oh, I had not gotten that impression. I thought this doc was intended to explain jj to people who have no VCS background.

I would call that a that doc a `Introduction to JJ`. To my understanding a from first principles doc tries to explain the core concepts and goals of the project.